src/events/upgrade.py

```
@@ -0,0 +1,100 @@
+
+    def post_upgrade_check(self) -> None:
+        """Runs necessary checks validating the unit is in a healthy state after upgrade."""
+        self.pre_upgrade_check()
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        pass
+
+    @override
+    def build_upgrade_stack(self) -> list[int]:
+        upgrade_stack = []
+        units = [self.charm.unit] + list(self.charm.state.peer_relation.units)
+        for unit in units:
+            upgrade_stack.append(int(unit.name.split("/")[-1]))
+
+        return upgrade_stack
+
+    @override
+    def log_rollback_instructions(self) -> None:
+        logger.critical(
+            "\n".join(
+                [
+                    "Unit failed to upgrade and requires manual rollback to previous stable version.",
+                    "    1. Re-run `pre-upgrade-check` action on the leader unit to enter 'recovery' state",
+                    "    2. Run `juju refresh` to the previously deployed charm revision",
+                ]
+            )
+        )
+        return
+
+    @override
+    def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
+        self.charm.workload.stop()
+
+        if not self.charm.workload.install():
+            logger.error("Unable to install OpensearchDashboards...")
+            self.set_unit_failed()
+            return
+
+        logger.info(f"{self.charm.unit.name} upgrading workload...")
+        self.charm.workload.restart()
+
+        time.sleep(5.0)
+
+        try:
+            logger.debug("Running post-upgrade check...")
+            self.post_upgrade_check()
+
+            logger.debug("Marking unit completed...")
+            self.set_unit_completed()
+
+            # ensures leader gets it's own relation-changed when it upgrades
+            if self.charm.unit.is_leader():
+                logger.debug("Re-emitting upgrade-changed on leader...")
+                self.on_upgrade_changed(event)
+
+        except ClusterNotReadyError as e:
```

When would this exception be raised? I assume on `post_upgrade_check`?

---

TBH in our case it's just to comply to the "usual template".

Normally `ClusterNotReadyError` is the one that's recommended to raise during the `pre_upgrade_check`.

So in case we add a `pre_upgrade_check` instead of the current empty one, that's when this could make sense.

---

Update: Having added a `pre_upgrade_check` and a `post_upgrade_check`, this catch is not only for "following protocols", but to take action on corresponding failure cases.

---

nit: `post_upgrade_check` should throw a different error, as the "cluster is ready".

---

OK, I could turn that one to a better one. I thnk I saw a version mismetch exception in the `upgades` lib.

Ill do the change within another PR (ast itls really small).